maykin_common/throttling.py

```diff
@@ -42,59 +81,94 @@ class ThrottleMixin:
+                "returning aHTTP 429 response instead.",
+                category=DeprecationWarning,
+                stacklevel=2,
+            )
+
+    def get_throttle_cache(self) -> BaseCache:
+        return caches[self.throttle_cache]
 
     def get_throttle_identifier(self) -> str:
-        user = getattr(self, "user_cache", self.request.user)
+        user = self.request.user
         return str(user.pk)
 
-    def create_throttle_key(self):
-        """
-        :rtype string Use as key to save the last access
+    def _get_throttle_window(self):
         """
+        Calculate the start of the window based on the current time.
 
-        return (
-            f"throttling_{self.get_throttle_identifier()}_"
-            f"{self.throttle_name}_{self.get_throttle_window()}"
-        )
+        This uses the current time (unix timestamp) as input and looks up the most
+        recent moment when a non-completed block of ``trottle_period`` started. It is
+        used as input in the cache key, meaning that once ``throttle_period`` has
+        elapsed, the throttle quota is fully reinstated.
 
-    def get_throttle_window(self):
-        """
-        round down to the throttle_period, which is then used to create the key.
+        Effectively, throttle intervals are fixed and not a sliding window.
         """
         current_time = int(time())
         return current_time - (current_time % self.throttle_period)
 
-    def get_visits_in_window(self):
+    def _get_num_visits_in_window(self) -> int:
         cache = self.get_throttle_cache()
-        key = self.create_throttle_key()
-
-        initial_visits = 1
-        stored = cache.add(key, initial_visits, self.throttle_period)
-        if stored:
-            visits = initial_visits
-        else:
-            try:
-                visits = cache.incr(key)
-            except ValueError:
-                visits = initial_visits
-        return visits
-
-    def should_be_throttled(self):
+        cache_key = (
+            f"throttling_{self.get_throttle_identifier()}_"
+            f"{self.throttle_name}_{self._get_throttle_window()}"
+        )
+
+        added = cache.add(cache_key, value=1, timeout=self.throttle_period)
+        if (
+            added
+        ):  # key added, we had no counter before -> one visit returned and stored
+            return 1
+
+        try:
+            return cache.incr(cache_key)
+        except ValueError:  # XXX: when does this happen?
+            return 1
+
+    def should_be_throttled(self) -> bool:
+        """
+        Determine if throttling is enabled for the request.
+        """
         if self.throttle_methods == "all":
             return True
         assert isinstance(self.request.method, str)
         return self.request.method.lower() in self.throttle_methods
 
-    def dispatch(self, request, *args, **kwargs):
+    @property
+    def rate_limit_exceeded(self) -> bool:
```

Comment:
@alextreme a new property you can check in your custom `dispatch` code _if you need to_ - I saw BD repeats the `if should_be_throttled and self.get_visits_in_window() ...` code in some derived mixins.

Though the proper way now is to not touch `dispatch`, set/keep `throttle_403 = False` and instead implement the method `handle_rate_limit_exceeded`.